[resharding] Generate intervals for non account_id prefix trie keys

shreyan-gupta opened the pull request:
We were not handling creating the intervals for keys like delayed receipts or buffered receipts.

This PR handles creating intervals for these other-category keys based on the resharding design.

Additionally simplify the resharding retain_split_shard interface.

shreyan-gupta commented on core/store/src/trie/ops/tests.rs:
```diff
@@ -0,0 +1,389 @@
+    (trie, memtries)
+}
+
+// fn retain_split_shard_custom_ranges<'a, N, V>(
```
Will remove in the first iteration of comment resolution.

Longarithm commented on core/store/src/trie/ops/resharding.rs:
```diff
@@ -214,3 +254,113 @@ fn retain_decision(key: &[u8], intervals: &[Range<Vec<u8>>]) -> RetainDecision {
     }
 }
+
+pub(crate) trait GenericTrieUpdateSplit<'a, N, V>:
```
It was an explicit decision to name these Retain (and the name below would be RetainInner) because this is what is being done. We don't split shard, we only retain relevant keys, especially given that some ranges are left for both shards.

shreyan-gupta replied:
Changed name to `Retain`

Longarithm reviewed (commented):
Why we are moving large amounts of code here?

Could you clarify more what is simplified by new traits? Intuitively I understand this but can't navigate by the big diff.

shreyan-gupta commented:
> Why we are moving large amounts of code here? It makes it harder to understand the change especially when it's not described.
> 
> Could you clarify more what is simplified by new traits? Intuitively I understand this but can't navigate by the big diff.

Apologies for that, this PR got a bit complicated as I was first adding things and then had to remove them. For simplicity, let me create a separate PR with the core logic and convert this one to a refactoring one while addressing your comments.

shreyan-gupta commented on core/store/src/trie/ops/resharding.rs:
```diff
@@ -33,27 +34,66 @@ enum RetainDecision {
+            _ => panic!("Unhandled trie key type: {}", name),
         }
     }
+    debug_assert!(intervals.iter().all(|range| range.start < range.end));
```
The intervals generated here are deterministic and always the same. It's possible to not have any keys within those intervals in the trie. Infact, this check can be pushed into the tests.

shreyan-gupta commented:
> > Why we are moving large amounts of code here? It makes it harder to understand the change especially when it's not described.
> > Could you clarify more what is simplified by new traits? Intuitively I understand this but can't navigate by the big diff.
> 
> Apologies for that, this PR got a bit complicated as I was first adding things and then had to remove them. For simplicity, let me create a separate PR with the core logic and convert this one to a refactoring one while addressing your comments.

Created a separate PR https://github.com/near/nearcore/pull/12521
